content/blog/upkeep-testthat-3/index.Rmd

```diff
@@ -0,0 +1,148 @@
+    particular upgrading to testthat edition 3.
+
+photo:
+  url: https://unsplash.com/photos/46juD4zY1XA
+  author: David Pisnoy
+
+# one of: "deep-dive", "learn", "package", "programming", "roundup", or "other"
+categories: [learn] 
+tags: [usethis, devtools]
+---
+
+<!--
+TODO:
+* [x] Look over / edit the post's title in the yaml
+* [x] Edit (or delete) the description; note this appears in the Twitter card
+* [x] Pick category and tags (see existing with `hugodown::tidy_show_meta()`)
+* [x] Find photo & update yaml metadata
+* [x] Create `thumbnail-sq.jpg`; height and width should be equal
+* [x] Create `thumbnail-wd.jpg`; width should be >5x height
+* [x] `hugodown::use_tidy_thumbnails()`
+* [ ] ~Add intro sentence, e.g. the standard tagline for the package~
+* [ ] ~`usethis::use_tidy_thanks()`~
+-->
+
+On the tidymodels team, I split time between new things like the [censored package for survival analysis](https://www.tidyverse.org/blog/2021/11/survival-analysis-parsnip-adjacent/) and keeping the existing packages in good shape.
+
+
+## Upkeep of tidy packages
+
+As the collection of packages grows, maintenance becomes increasingly important. Hadley Wickham made it the topic of his [keynote at rstudio::global 2021](https://www.rstudio.com/resources/rstudioglobal-2021/maintaining-the-house-the-tidyverse-built/). 
+The [usethis](https://usethis.r-lib.org/) package provides tools for the workflows when working on tidyverse, r-lib, or tidymodels packages and as such contains the set of labels used to categorize GitHub issues. Day-to-day maintenance is captured via the `"upkeep"` label for "maintenance, infrastructure, and similar". For bigger efforts, `use_tidy_upkeep_issue()` creates an issue containing a checklist of actions to bring a package up to current tidyverse standards.
+
+![A GitHub issue with a checklist of upkeep tasks](upkeep-issue.png)
+
+
+## testthat 3e
+ 
+One of the items on this checklist is to use testthat edition 3. The testthat package introduced the idea of an "edition" in version 3.0.0:
+
+> An edition is a bundle of behaviours that you have to explicitly choose to use, allowing us to make otherwise backward incompatible changes.
+
+While you can continue to use testthat's previous behaviour, it is recommended to upgrade so that you can make use of handy new features. As some of the changes may break your tests, you might have been putting that off though? You would not be alone in that! Several tidymodels packages still have to make the jump but I recently upgraded [dials](https://github.com/tidymodels/dials/) and [censored](https://github.com/tidymodels/censored/) to testthat edition 3. Here is what I did and learned along the way.
+
+### Workflow to upgrade
+
+First off, the [testthat article introducing the 3rd edition](https://testthat.r-lib.org/articles/third-edition.html) is very helpful. I'd recommend reading that if you haven't yet!
```

Comment:
I'd mention this in the previous section as a way for folks to learn more about 3e, if they haven't heard about it before.